test/e2e/e2e_test.go

```diff
@@ -87,6 +89,109 @@ var _ = Describe("E2E", func() {
 				// TODO Check that the expected CRDs, deployments, services, etc. are present
 			})
 		})
+		Context("Mixerless Telemetry Stats Filter Test", func() {
+			// var resourcesFile string
+			var (
+				istio        v1beta1.Istio
+				filterBefore string
+				err          error
+				statsname    string
+				tcpname      string
+			)
+
+			BeforeEach(func() {
+				//
+			})
+
+			JustBeforeEach(func() {
+				// get the istio object created in the OUTER JustBeforeEach
+				log.Info("Namespace: ", "Namespace", instance.Namespace)
+				log.Info("Name: ", "Name", instance.Name)
+				filterBefore, err = GetMixerlessTelemetryStatus(&istio, instance.Namespace, instance.Name)
+				Expect(err).NotTo(HaveOccurred())
+				log.Info("JustBeforeEach: ", "filterBefore", filterBefore)
+				version := istio.Spec.Version // get first 2 digits
+				verParts := strings.SplitN(string(version), ".", 3)
+				majMinor := fmt.Sprintf("%s.%s", verParts[0], verParts[1])
+				log.Info("Istio Version: ", "Version", majMinor)
+
+				// Names of the filters of interest
+				statsname = istio.WithRevision(fmt.Sprintf("mixerless-telemetry-stats-filter-%s", majMinor))
+				tcpname = istio.WithRevision(fmt.Sprintf("mixerless-telemetry-tcp-stats-filter-%s", majMinor))
+				log.Info("Filter Names: ", "Stats", statsname, "TCP", tcpname)
+
+			})
+
+			AfterEach(func() {
+				var expectMissingFilter bool
+				if filterBefore != "" {
+					filterNow, err := GetMixerlessTelemetryStatus(&istio, instance.Namespace, instance.Name)
+					Expect(err).NotTo(HaveOccurred())
+					if filterNow != filterBefore {
+						log.Info("AfterEach: Restore", "filterBefore", filterBefore, "filterNow", filterNow)
+						expectMissingFilter, err = SetMixerlessTelemetryState(
+							&istio, instance.Namespace, instance.Name, filterBefore)
+						Expect(err).NotTo(HaveOccurred())
+						err = WaitForMixerlessTelemetryFilter(
+							istio.Namespace, statsname, expectMissingFilter, 300*time.Second, 5*time.Second)
+						Expect(err).NotTo(HaveOccurred())
+
+						err = WaitForMixerlessTelemetryFilter(
+							istio.Namespace, tcpname, expectMissingFilter, 300*time.Second, 5*time.Second)
+						Expect(err).NotTo(HaveOccurred())
+					}
+				}
+			})
+
+			It("Transitions Filter through all states", func() {
```

Comment:
I think it would be better to avoid the `stateTransitions` variable and simply write down all steps. Also, it's enough to store the "enabled" state in the setup and reset it in teardown. The test itself doesn't have to depend on the initial state.

I'm thinking something like:

```
It("...", func() {
	SetMixerlessTelemetryState(istio, true)
	Expect(WaitForMixerlessTelemetryEnvoyFilters(istio, pkgutil.PointerToBool(true), 15*Seconds, 100*Millis)).To(Succeed())

	SetMixerlessTelemetryState(istio, false)
	Expect(WaitForMixerlessTelemetryEnvoyFilters(istio, pkgutil.PointerToBool(false), 15*Seconds, 100*Millis)).To(Succeed())

	SetMixerlessTelemetryState(istio, true)
	Expect(WaitForMixerlessTelemetryEnvoyFilters(istio, pkgutil.PointerToBool(true), 15*Seconds, 100*Millis)).To(Succeed())

	SetMixerlessTelemetryState(istio, nil)
	Expect(WaitForMixerlessTelemetryEnvoyFilters(istio, pkgutil.PointerToBool(nil), 15*Seconds, 100*Millis)).To(Succeed())
})
```

The signature of `WaitForMixerlessTelemetryEnvoyFilters` would be something like:
```
func WaitForMixerlessTelemetryEnvoyFilters(istio *Istio, shouldExist bool, timeout Duration, interval Duration) err
```

You can sprinkle it with a couple `Sleep()` calls, although I don't think it's necessary here.

Reply:
No sleeps were needed since we poll for the change.

Reply:
modified to use filterShouldExist and added Expect().Should(Succeed())